### PR TITLE
WIP: SOSS with XTypes

### PR DESCRIPTION
**Pull request by lemunozm**
In this PR we added the *xtypes* functionality to *soss*.

Things to take into account:
* We added another package as an example of how to build a system handle (especially focused on the use of *soss* with *xtypes*). The idea is to let this package grow with ever new features coming from *xtypes* module, and if wanted, keep it clean allowing the user to take it as a template for other system handles.

Currently the following packages have been updated:
- Internal packages:
  - [x] core
  - [x] mock
  - [x] echo (formerly called local-example)
  - [x] ros2 and ros2-test 
  - [x] json
  - [x] websocket and websocket-test.
- External packages (from osrf):
  - [ ] ros1 and ros1-test
- External packages (from eProsima):
  - [x] fiware and fiware-test (https://github.com/eProsima/SOSS-FIWARE/pull/17)
  - [x] dds and dds-test (https://github.com/eProsima/SOSS-DDS/pull/24)
- Other features
  - [x] YAML configuration file with IDL. 

Please, feel free to give us feedback about the incorporation of *xtypes* to *soss*.
I will update this PR description with links to the documentation and examples:
- [xtypes API documentation](https://github.com/eProsima/xtypes)
- [xtypes API usage example](https://github.com/eProsima/xtypes/blob/master/examples/complex_type.cpp)
- [System Handle conversion code example](https://github.com/eProsima/soss_v2/blob/feature/xtypes-dds/packages/json/src/conversion.cpp) (updated)
- [Functional QoS round trip example: Vec2 -> Vec3 -> Vec3 -> Vec2](https://github.com/eProsima/soss_v2/tree/feature/xtypes-dds/packages/echo/example)

### CODE REVIEW

**Comment by marcoag** [comment]
It seems that this brings a dependency on dds (`#include <dds/core/xtypes/xtypes.hpp>`). I understand the final objective of this is to create an independent package for xtypes. If that is the case how about we merge this into a branch (i.e. soss_v3) in the meantime until the packaging work is completed?

**Comment by lemunozm** [comment]
> It seems that this brings a dependency on dds (`#include <dds/core/xtypes/xtypes.hpp>`). I understand the final objective of this is to create an independent package for xtypes. If that is the case how about we merge this into a branch (i.e. soss_v3) in the meantime until the packaging work is completed?

Hi @marcoag, 

At this moment, the dependency is managed as a thirdparty. When we reach a stable version, this dependency could be managed as an external project (as `yamlcpp` is managed currently in `soss_v2`). 
Anyway, we are open to new ideas.

File: .gitmodules
Comment (marcoag) [inline .gitmodules]: afaik git submodules tend to be error prone as they require some good git knowledge and discipline, specially when trying to keep versions between branches synced, any alternative to this we can think of?

File: doc/soss-xtypes-upgrade.md
Comment (marcoag) [inline doc/soss-xtypes-upgrade.md:138]: This applies to all the "previous"  on the md file:
 
Do we really need to keep the "previous" way of doing it? is it still supported?

File: packages/core/include/soss/Instance.hpp
Comment (marcoag) [inline packages/core/include/soss/Instance.hpp]: exit or exist?

File: .gitmodules
Comment (lemunozm) [reply .gitmodules]: That's true, the actual use of *git submodules* is temporal. We are analysing the idea to manage these dependencies with the *superbuild* pattern of `cmake`.

Comment (marcoag) [reply .gitmodules]: I see... is all the code in those three repos needed for the integration of xtypes? would it be an option to somehow extract the relevant parts of some specific stable version and move them into thirdparty?

File: packages/core/include/soss/Message.hpp
Comment (marcoag) [inline packages/core/include/soss/Message.hpp]: Is this file still needed or can we move this include to the relevant files?

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]: Could we find a better way than hardcoding the "mock" middleware option here?

**Review by marcoag — COMMENTED** [review]
After updating the git submodules I get this when trying to colcon build:

` fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory`

What would be the process to build SOSS after this PR is merged?

File: doc/soss-xtypes-upgrade.md
Comment (lemunozm) [reply doc/soss-xtypes-upgrade.md:138]: The "previous" way is not supported. This document is only to explain the soss transition to xtypes, or to help some user to adapt its SystemHandle for working with xtypes.

File: packages/core/include/soss/Message.hpp
Comment (lemunozm) [reply packages/core/include/soss/Message.hpp]: The purpose of `dds/core/xtypes/xtypes.hpp` is to bring all the necessary of xtypes to the soss ecosystem.

I do not know if you are referring to make separated *includes*, (one for each type for example).  If this is the case, the aim is to simplify the user management: if he writes the `xtypes` namespace, he has access to the public interface of the module. We have chosen this strategy because all xtypes entities are very coupled.

File: packages/core/src/Config.cpp
Comment (lemunozm) [reply packages/core/src/Config.cpp]: I will investigate, maybe the own `mock` SH can check itself for this property.

File: packages/core/include/soss/Instance.hpp
Comment (lemunozm) [reply packages/core/include/soss/Instance.hpp]: You are right, I will modify it.

**Comment by lemunozm** [comment]
> After updating the git submodules I get this when trying to colcon build:
> 
> ` fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory`
> 
> What would be the process to build SOSS after this PR is merged?

Hi @marcoag ,
The following is enough for me (and avoid to compile packages no updated with xtypes): 
```
colcon build --packages-up-to  soss-local-example
```

When you get the error: ` fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory`, what package is complining?

File: .gitmodules
Comment (lemunozm) [reply .gitmodules]: *xtypes* could be separated in a future. At this moment, we have the library inside because we are developing `fastrtps` over it, and xtypes is an integral part of it. An advantage of it is that as `ros2` use `fastrtps` inside, the `soss-ros2` system handle could avoid the conversion phase because its native message can propagate to soss directly. The same can be applied to `soss-dds` system handle.

Anyway, can be a topic of discussion for the next meeting.

Comment (marcoag) [reply .gitmodules]: I believe if we could have separate code for `xtypes` would be the best as `soss` is not (and shouldn't be) necessarily bounded to`ros2`, so it's dependency on `fastrtps` just for `xtypes` becomes a bit of an overkill solution imho.

File: doc/soss-xtypes-upgrade.md
Comment (marcoag) [reply doc/soss-xtypes-upgrade.md:138]: Ok, let's keep it for now as a reference then

File: packages/core/src/Config.cpp
Comment (marcoag) [reply packages/core/src/Config.cpp]: I think that would look better, yes.

File: packages/core/include/soss/Message.hpp
Comment (marcoag) [reply packages/core/include/soss/Message.hpp]: I was referring to Message.hpp, as it currently contains an include... maybe just change all includes that point to Message.hpp to xtypes.hpp and remove Message.hpp?

**Comment by marcoag** [comment]
Maybe because you have fastrtps installed on your system? (I understand this shouldn't be a requirement for compiling soss)

```
marco@marco:~/workspaces$ colcon build --packages-up-to soss-local-example
Starting >>> fastcdr 
Starting >>> foonathan_memory_vendor
Finished <<< fastcdr [0.23s]                                                                                 
Finished <<< foonathan_memory_vendor [0.24s]
Starting >>> fastrtps
Finished <<< fastrtps [0.26s]                       
Starting >>> soss-core
--- stderr: soss-core                             
In file included from /home/marco/workspaces/soss/src/soss/packages/core/include/soss/SystemHandle.hpp:23:0,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/register_system.hpp:21,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/register_system.cpp:18:
/home/marco/workspaces/soss/src/soss/packages/core/include/soss/Message.hpp:21:10: fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory
 #include <dds/core/xtypes/xtypes.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/soss-core.dir/src/register_system.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /home/marco/workspaces/soss/src/soss/packages/core/include/soss/SystemHandle.hpp:23:0,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/register_system.hpp:21,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/Config.hpp:21,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/Instance.cpp:18:
/home/marco/workspaces/soss/src/soss/packages/core/include/soss/Message.hpp:21:10: fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory
 #include <dds/core/xtypes/xtypes.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/soss-core.dir/src/Instance.cpp.o] Error 1
In file included from /home/marco/workspaces/soss/src/soss/packages/core/include/soss/SystemHandle.hpp:23:0,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/register_system.hpp:21,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/Config.hpp:21,
                 from /home/marco/workspaces/soss/src/soss/packages/core/src/Config.cpp:19:
/home/marco/workspaces/soss/src/soss/packages/core/include/soss/Message.hpp:21:10: fatal error: dds/core/xtypes/xtypes.hpp: No such file or directory
 #include <dds/core/xtypes/xtypes.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/soss-core.dir/src/Config.cpp.o] Error 1
make[1]: *** [CMakeFiles/soss-core.dir/all] Error 2
make: *** [all] Error 2
---
Failed   <<< soss-core	[ Exited with code 2 ]

Summary: 3 packages finished [0.77s]
  1 package failed: soss-core
  1 package had stderr output: soss-core
  2 packages not processed
```

File: .gitmodules
Comment (lemunozm) [reply .gitmodules]: *xtypes* is mainly inside of *fastrtps* because from the *fastrtps* view it is a core part and the  *fastrtps* development is easier with this module inside. Anyway, we understand perfectly the *soss* view in this regard. We are currently working in to extract *xtypes* from *fastrtps* to offer this feature without their dependencies. I hope in the next few days to offer a SOSS commit only with *xtypes* dependecy.

File: packages/core/include/soss/Message.hpp
Comment (lemunozm) [reply packages/core/include/soss/Message.hpp]: I left the `Message.hpp` file there in order to pack some others inclusions that the messaging might need, or maybe some future utility over the `xtypes` that `soss-core` could offer. The user only needs to carry about the `Message.hpp` inclusion.

**Comment by lemunozm** [comment]
> colcon build --packages-up-to soss-local-example

We updated `xtypes` with some `cmake` modifications the last week. Are you testing with the last commit?

**Comment by lemunozm** [comment]
Could you update your `soss_v3_devel` branch to `master`? We are developing our `xtypes-dds` branch over some features that `master` branch offers.

File: .gitmodules
Comment (marcoag) [reply .gitmodules]: That sounds great, thanks!

**Comment by mxgrey** [comment]
It seems that even with the target branch completely up to date with master, git still considers some of the files to be in conflict. I'm not sure what can be done besides manually resolving the merge conflicts.

**Comment by mxgrey** [comment]
Maybe if you merge the current `master` branch into your remote `eProsima/soss_v2:feature/xtypes-dds` branch, that could automatically resolve the conflicts?

**Comment by lemunozm** [comment]
> Maybe if you merge the current `master` branch into your remote `eProsima/soss_v2:feature/xtypes-dds` branch, that could automatically resolve the conflicts?

Hi Grey,
yes, we were waiting for your rebase. 
Out branch is already rebased too.
Thanks.

File: packages/core/include/soss/Message.hpp
Comment (marcoag) [reply packages/core/include/soss/Message.hpp]: ok, we can keep it like that then.

**Comment by marcoag** [comment]
It seems that `colcon build --packages-up-to  soss-local-example` works now, however when building the rest `soss-ros2` still shows the error pointing out that the `xtypes.hpp` file is not found, maybe we should update the packages that need this so we don't break the overall build.

**Comment by lemunozm** [comment]
> It seems that `colcon build --packages-up-to soss-local-example` works now, however when building the rest `soss-ros2` still shows the error pointing out that the `xtypes.hpp` file is not found, maybe we should update the packages that need this so we don't break the overall build.

At this moment, `soss-ros2` is not ported to *xtypes* (the same happens with `soss-websocket` package). We are working now into adding *xtypes* to `soss-ros2`.

File: .gitmodules
Comment (lemunozm) [reply .gitmodules]: The newly added commit uses *xtypes* as the only third-party. In the future, and in order to avoid git management of submodules, we can add it as a *CMake ExternalProject* and download it automatically.

File: packages/core/src/Config.cpp
Comment (lemunozm) [reply packages/core/src/Config.cpp]: Already modified.

File: .gitmodules
Comment (marcoag) [reply .gitmodules]: Yup, that sounds like a good idea.

File: .gitmodules
Comment (marcoag) [inline .gitmodules:3]: Might be more robust to have this dependency as a CMake External Project and download it automatically.

File: packages/core/include/soss/Message.hpp
Comment (marcoag) [inline packages/core/include/soss/Message.hpp:26]: Is there a need for this file anymore? 

In the same line, Messages.cpp might not be necessary either...

File: packages/core/CMakeLists.txt
Comment (marcoag) [inline packages/core/CMakeLists.txt:58]:
```suggestion
```
Can I assume we don't need this?

File: packages/core/include/soss/SystemHandle.hpp
Comment (marcoag) [inline packages/core/include/soss/SystemHandle.hpp]:
```suggestion
  /// \param[in] type_registry
```
Although maybe type_registry a more specific name for this parameter can add more value

File: packages/echo/example/main.cpp
Comment (marcoag) [inline packages/echo/example/main.cpp:11]: maybe better not to hardcode paths, shall we put the config file as a parameter?

File: packages/core/src/MiddlewareInterfaceExtension.cpp
Comment (marcoag) [inline packages/core/src/MiddlewareInterfaceExtension.cpp:30]:
```suggestion
```

I'm not sure if this include is needed, on Linux is not needed for a successful colcon compilation. In any case either this or the one on line 26 but not both I guess.

File: packages/mock/CMakeLists.txt
Comment (marcoag) [inline packages/mock/CMakeLists.txt:5]: Doesn't seem to be needed, however maybe this line should remain if this mock plugin is supposed to serve as an example... shouldn't it?

**Comment by marcoag** [comment]
Branch has been updated to master and there seems to be some conflicts, @lemunozm could you guys resolve this on your side?

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
    std::cerr << "The config file [" << filename << "] has a 'types' entry "
              << "which doesn't contains an 'idls' entry."
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
          std::cout << "Parsing the idl '" << idl_index << "' placed in the "
                    << "yaml config. The parsing was successful but no "
                    <<"types were found."
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
          std::cerr << "Error parsing the idl '" << idl_index << "' placed in "
                << "the yaml config. Please, review and fix your idl "
                << "specification syntax."
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
        std::cerr << channel_type << " configuration [" << name
                  << "] is missing its [type], or [request_type] and "
                  << "[reply_type] fields!" << std::endl;
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
            std::cerr << "Unrecognized system [" << it_remap.first
                  << "] requested for mapping topic ["
                  << entry.first << "]" << std::endl;
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
        std::cerr << "Unrecognized system [" << it_remap.first 
                  << "] requested for mapping service ["
                  << entry.first << "]" << std::endl;
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
         std::cerr << "'types-from' references to a non-existant middleware"
                   << std::endl;
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
          std::cerr << "The middleware '" << mw_name << "' must satisfy the "
                    << "required type '" << required_type << "'"
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
          std::cerr << "The middleware '" << mw_name << "' must satisfy the "
                    << "required type '" << required_type << "'"
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
    TopicInfo server_info = remap_if_needed(server, 
                                            config.remap, 
                                            {service_name, 
                                             config.request_type, 
                                             config.reply_type});
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
```
maybe?

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
      const xtypes::DynamicType* server_reply_type = 
          resolve_type(it->second.types, 
                       server_info.reply_type);
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
```
maybe?

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
```
maybe?

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
```
?

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion
            else //previously ensured that TypeConsistency is not NONE
```

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion

    for (const std::string& to : config.route.to)
    {
      const auto it_to = info_map.find(to);
      TopicInfo topic_info_to = remap_if_needed(to, 
                                                config.remap, 
                                                {topic_name, 
                                                 config.message_type, ""});
      const xtypes::DynamicType* to_type = resolve_type(it_from->second.types,
                                                        topic_info_from.type);
```
Shall we remove the commented code?

File: packages/core/src/Config.cpp
Comment (marcoag) [inline packages/core/src/Config.cpp]:
```suggestion

    const auto it_server = info_map.find(config.route.server);
    TopicInfo topic_info_server = remap_if_needed(config.route.server, 
                                                  config.remap,
                                                  {service_name, 
                                                   config.request_type,
                                                   config.reply_type});
    const xtypes::DynamicType* server_type = 
        resolve_type(it_server->second.types, topic_info_server.type);
```
Shall we remove the commented code?

File: packages/ros2-test/integration/ros2__test_msgs.cpp
Comment (marcoag) [inline packages/ros2-test/integration/ros2__test_msgs.cpp:260]:
```suggestion
        CHECK(received_array.primitive_values[i] 
                == ros2_msg.primitive_values[i]);
```

**Review by marcoag — CHANGES_REQUESTED** [review]

**Comment by russkel** [comment]
Hi, is this still under development? I am quite interested in the xtypes features for a project.

**Comment by mxgrey** [comment]
Due to a shortage of resources that Open Robotics can commit to SOSS development, I believe eProsima intends to take over as the maintainers of the SOSS/Integration-Service project. Perhaps @lemunozm could advise on where to find the latest developments.

**Comment by lemunozm** [comment]
Hi @russkel, @mxgrey, I am no longer working on this project, sorry, I can't help

**Comment by russkel** [comment]
Ah that's unfortunate this was never merged in. Thanks for the update.

**Comment by jamoralp** [comment]
> Ah that's unfortunate this was never merged in. Thanks for the update.

Good morning @russkel, as part of the team in eProsima responsible of this project I am going to be mainly involved in transferring the repository to our organization and updating the main branch with the xTypes implementation. Keep an eye on it, it should be available in the following months. I will keep you updated!

**Comment by russkel** [comment]
Hi @jamoralp, I saw the new commits to the repository and was relieved! Thanks for the message and your work on it.

**Comment by jamoralp** [comment]
New SOSS_v3 release will be named *Integration Service* and will feature the whole set of functionalities added in this pull request regarding `xtypes`. The `feature/xtypes-dds` branch will be merged onto a brand new `main` branch by april. Therefore, we can close this

**Pull request closed without merging**
